core/src/test/scala/unit/kafka/server/EdgeCaseRequestTest.scala

```diff
@@ -0,0 +1,181 @@
+        outgoing.writeInt(request.length + 2)
+        outgoing.writeShort(id)
+      case None =>
+        outgoing.writeInt(request.length)
+    }
+    outgoing.write(request)
+    outgoing.flush()
+  }
+
+  private def receiveResponse(socket: Socket): Array[Byte] = {
+    socket.isInputShutdown
+    val incoming = new DataInputStream(socket.getInputStream)
+    val len = incoming.readInt()
+    val response = new Array[Byte](len)
+    incoming.readFully(response)
+    response
+  }
+
+  private def requestHeaderBytes(apiKey: Short, apiVersion: Short, clientId: String = "", correlationId: Int = -1): Array[Byte] = {
+    val header = new RequestHeader(apiKey, apiVersion, clientId, correlationId)
+    val buffer = ByteBuffer.allocate(header.sizeOf())
+    header.writeTo(buffer)
+    buffer.array()
+  }
+
+  @Test
+  def testProduceRequestWithNullClientId() {
+    val topic = "topic"
+    val topicPartition = new TopicPartition(topic, 0)
+    val correlationId = -1
+    TestUtils.createTopic(zkUtils, topic, numPartitions = 1, replicationFactor = 1, servers = Seq(server))
+
+    val serializedBytes = {
+      val headerBytes = requestHeaderBytes(ApiKeys.PRODUCE.id, 1, null, correlationId)
+      val messageBytes = "message".getBytes
+      val request = new ProduceRequest(1, 10000, Map(topicPartition -> ByteBuffer.wrap(messageBytes)).asJava)
+      val byteBuffer = ByteBuffer.allocate(headerBytes.length + request.sizeOf)
+      byteBuffer.put(headerBytes)
+      request.writeTo(byteBuffer)
+      byteBuffer.array()
+    }
+
+    val plainSocket = connect(protocol = SecurityProtocol.PLAINTEXT)
+    sendRequest(plainSocket, serializedBytes)
+    val response = receiveResponse(plainSocket)
+    plainSocket.close()
+
+    val responseBuffer = ByteBuffer.wrap(response)
+    val responseHeader = ResponseHeader.parse(responseBuffer)
+    val produceResponse = ProduceResponse.parse(responseBuffer)
+
+    assertEquals("The response should parse completely", 0, responseBuffer.remaining())
+    assertEquals("The correlationId should match request", correlationId, responseHeader.correlationId())
+    assertEquals("One partition response should be returned", 1, produceResponse.responses().size())
+
+    val partitionResponse = produceResponse.responses().get(topicPartition)
+    assertNotNull(partitionResponse)
+    assertEquals("There should be no error", 0, partitionResponse.errorCode)
+  }
+
+  @Test
+  def testHeaderOnlyRequest() {
+    val plainSocket = connect(protocol = SecurityProtocol.PLAINTEXT)
+    sendRequest(plainSocket, requestHeaderBytes(ApiKeys.PRODUCE.id, 1))
+
+    assertEquals("The server should disconnect", -1, plainSocket.getInputStream.read())
+    plainSocket.close()
+  }
+
+  @Test
+  def testInvalidApiKeyRequest() {
+    val plainSocket = connect(protocol = SecurityProtocol.PLAINTEXT)
+    sendRequest(plainSocket, requestHeaderBytes(-1, 0))
+
+    assertEquals("The server should disconnect", -1, plainSocket.getInputStream.read())
```

Comment:
Seems like it's a common pattern to connect, send request, get input stream and read it and then close the socket (ideally in a finally block). Maybe extract a method?